File: test_instances/catch2_unit_tests/backend_testbed.cpp
#include "backend_testbed.h"
#include <catch2/catch_test_macros.hpp>

// Global flag useCatch2 switches Catch2 on or off.
bool useCatch2 = true;

void disableCatch2() {
    useCatch2 = false;
}

TreeNode createNoContentTreeNode(string label_rule, string description, vector<TreeNode::PropertyInfo> property_infos, 
    TreeNodeVersion version, vector<string> child_names, 
    maybe<string> query_how_to, maybe<string> qa_sequence) {
    shared_span<> no_content(global_no_chunk_header, false);
    return TreeNode(label_rule, description, property_infos, version, child_names, std::move(no_content), std::move(query_how_to), std::move(qa_sequence));
}

TreeNode createAnimalNode(string animal, string description, vector<TreeNode::PropertyInfo> property_infos, 
    TreeNodeVersion version, vector<string> child_names, 
    vector<string> property_data, string query_how_to, string qa_sequence) {
    payload_chunk_header header(0, payload_chunk_header::SIGNAL_WWATP_UPSERT_NODE_REQUEST, 0);
    shared_span<> animal_data(header, true);
    pair<bool, pair<size_t, size_t>> next = {false, {0, 0}};
    uint16_t index = 0;
    for (auto& content : property_data) {
        auto info = property_infos[index];
        if(info.first == "int") {
            stringstream ss(content);
            uint64_t value;
            ss >> value;
            next = {true, animal_data.copy_type<uint64_t>(value, next)};
        }
        if(info.first == "double") {
            stringstream ss(content);
            double value;
            ss >> value;
            next = {true, animal_data.copy_type<double>(value, next)};
        }
        if(info.first == "float") {
            stringstream ss(content);
            float value;
            ss >> value;
            next = {true, animal_data.copy_type<float>(value, next)};
        
        }
        if(info.first == "bool") {
            stringstream ss(content);
            bool value;
            ss >> std::boolalpha >> value;
            next = {true, animal_data.copy_type<bool>(value, next)};
        }
        if (fixed_size_types.find(info.first) == fixed_size_types.end()) {
            next.second = animal_data.copy_type<uint64_t>(content.size(), next);
            next.second = animal_data.copy_span<const char>(std::span<const char>(content.c_str(), content.size()), next);
        }
        index++;
    }
    auto only_animal_data = animal_data.restrict_upto(next.second);
    auto data_size = only_animal_data.size();
    only_animal_data.get_signal<payload_chunk_header>().data_length = static_cast<uint16_t>(data_size);
    if (property_data.empty()) {
        only_animal_data = shared_span<>(global_no_chunk_header, false);
    }
    auto m_query_how_to = maybe<string>(query_how_to);
    if (useCatch2) {
        REQUIRE(m_query_how_to.get_with_default("") == query_how_to);
    }
    else {
        assert(m_query_how_to.get_with_default("") == query_how_to);
    }
    auto m_qa_sequence = maybe<string>(qa_sequence);
    if (useCatch2) {
        REQUIRE(m_qa_sequence.get_with_default("") == qa_sequence);
    }
    else {
        assert(m_qa_sequence.get_with_default("") == qa_sequence);
    }
    return TreeNode(animal, description, property_infos, version, child_names, std::move(only_animal_data), m_query_how_to, m_qa_sequence);
}

vector<TreeNode> createAnimalDossiers(TreeNode &animal_node) {
    vector<TreeNode> animal_dossiers;
    for (const auto& child_name : animal_node.getChildNames()) {
        string child_label_rule = animal_node.getLabelRule() + "/" + child_name;
        TreeNodeVersion child_version = {1, 0, "public", maybe<string>(), maybe<string>(), maybe<string>(), maybe<int>(2)};
        TreeNode child_node = createNoContentTreeNode(child_label_rule, "Animal Dossier", {}, child_version, {}, maybe<string>(), maybe<string>());
        vector<string> note_names;
        for (int i = 0; i < 3; ++i) {
            string note_label_rule = child_label_rule + "/note" + std::to_string(i);
            note_names.push_back("note" + std::to_string(i));
            TreeNodeVersion note_version = {1, 0, "public", maybe<string>(), maybe<string>(), maybe<string>(), maybe<int>(2)};
            TreeNode note_node = createNoContentTreeNode(note_label_rule, child_name + " note " + std::to_string(i), {}, note_version, {}, maybe<string>(), maybe<string>());
            animal_dossiers.push_back(std::move(note_node));
        }
        child_node.setChildNames(note_names);
        animal_dossiers.push_back(std::move(child_node));
    }
    return animal_dossiers;
}

vector<TreeNode> createLionNodes() {
    TreeNode lion = createAnimalNode(
        "lion",
        "King of the jungle", 
        {{"int", "popularity"}, {"string", "diet"}},
        {1, 0, "public", maybe<string>(), maybe<string>("tester"), maybe<string>("tester"), maybe<int>(2)}, 
        {"Simba", "Nala"},
        {"10", "carnivore"},
        "url duh!", 
        "Zookeeper1: Lions are majestic.\nZookeeper2: Indeed, they are the kings of the jungle."
    );
    vector<TreeNode> lion_nodes = createAnimalDossiers(lion);
    lion_nodes.insert(lion_nodes.begin(), lion);
    return lion_nodes;
}

vector<TreeNode> createElephantNodes() {
    TreeNode elephant = createAnimalNode(
        "elephant", 
        "Largest land animal", 
        {{"int", "popularity"}, {"string", "diet"}},
        {1, 0, "public", maybe<string>(), maybe<string>(), maybe<string>(), maybe<int>(2)}, 
        {"Dumbo", "Babar"}, 
        {"8", "herbivore"},
        "url duh!", 
        "Zookeeper1: Elephants are so strong.\nZookeeper2: And they have great memory!"
    );
    vector<TreeNode> elephant_nodes = createAnimalDossiers(elephant);
    elephant_nodes.insert(elephant_nodes.begin(), elephant);
    if (elephant.getQueryHowTo().get_with_default("") != "url duh!") {
        std::cerr << "Elephant node has incorrect qa_sequence" << std::endl;
        throw std::runtime_error("Elephant node has incorrect qa_sequence");
    }
    return elephant_nodes;
}

vector<TreeNode> createParrotNodes() {
    TreeNode parrot = createAnimalNode(
        "parrot", 
        "Colorful bird", 
        {{"int", "popularity"}, {"string", "diet"}},
        {1, 0, "public", maybe<string>(), maybe<string>(), maybe<string>(), maybe<int>(2)}, 
        {"Polly", "Jerome"}, 
        {"7", "omnivore"},
        "url duh!", 
        "Zookeeper1: Parrots can mimic sounds.\nZookeeper2: Yes, they are very intelligent birds."
    );
    vector<TreeNode> parrot_nodes = createAnimalDossiers(parrot);
    parrot_nodes.insert(parrot_nodes.begin(), parrot);
    return parrot_nodes;
}

void checkGetNode(Backend const &backend, const string& label_rule, TreeNode const &expected_node) {
    maybe<TreeNode> node = backend.getNode(label_rule);
    if (node.is_just()) {
        auto found_node = node.get_with_default(TreeNode());
        if (useCatch2) {
            REQUIRE(found_node.getLabelRule() == expected_node.getLabelRule());
            REQUIRE(found_node == expected_node);
        }
        else {
            assert(found_node.getLabelRule() == expected_node.getLabelRule());
            assert(found_node == expected_node);
        }
    } else {
        if (useCatch2)
        {
            FAIL("Node not found: " + label_rule);
        }
        else
        {
            throw std::runtime_error("Node not found: " + label_rule);
        }
    }
}

void checkMultipleGetNode(Backend const &backend, const vector<TreeNode>& expected_nodes) {
    for (auto& expected_node : expected_nodes) {
        checkGetNode(backend, expected_node.getLabelRule(), expected_node);
    }
}

void checkDeletedNode(Backend const &backend, const string& label_rule) {
    auto node = backend.getNode(label_rule);
    if (useCatch2) {
        REQUIRE(node.is_nothing());
    }
    else {
        assert(node.is_nothing());
    }
}

void checkMultipleDeletedNode(Backend const &backend, const vector<TreeNode>& expected_nodes) {
    for (auto& expected_node : expected_nodes) {
        checkDeletedNode(backend, expected_node.getLabelRule());
    }
}

vector<TreeNode> collectAllNotes() {
    auto lion_nodes = createLionNodes();
    auto elephant_nodes = createElephantNodes();
    auto parrot_nodes = createParrotNodes();
    vector<TreeNode> all_notes;
    for (auto& lion_node : lion_nodes) {
        if (lion_node.getLabelRule().find("note") != string::npos) {
            all_notes.push_back(lion_node);
        }
    }
    for (auto& elephant_node : elephant_nodes) {
        if (elephant_node.getLabelRule().find("note") != string::npos) {
            all_notes.push_back(elephant_node);
        }
    }
    for (auto& parrot_node : parrot_nodes) {
        if (parrot_node.getLabelRule().find("note") != string::npos) {
            all_notes.push_back(parrot_node);
        }
    }
    return all_notes;
}

TreeNode createNotesPageTree() {
    auto noteNodes = collectAllNotes();
    vector<string> note_labels;
    for (auto& note : noteNodes) {
        note_labels.push_back(note.getLabelRule());
    }
    TreeNode notes_page_tree = createNoContentTreeNode("notes", "Animal Notes Page", {}, 
        {1, 0, "public", maybe<string>(), maybe<string>(), maybe<string>(), maybe<int>(2)}, 
        note_labels, maybe<string>(), maybe<string>());
    return notes_page_tree;
}

vector<TreeNode> prefixNodeLabels(string label_prefix, vector<TreeNode> nodes) {
    for (auto& node : nodes) {
        node.setLabelRule(label_prefix + node.getLabelRule());
        auto child_names = node.getChildNames();
        for (auto& child_name : child_names) {
            child_name = label_prefix + child_name;
        }
        node.setChildNames(child_names);
    }
    return nodes;
}

BackendTestbed::BackendTestbed(Backend& backend, bool should_test_notifications, bool should_test_changes)
    : backend_(backend), should_test_notifications_(should_test_notifications), 
      should_test_changes_(should_test_changes) {}

void BackendTestbed::stressTestConstructions(size_t count) {
    for(size_t i = 0; i < count; ++i) {
        addAnimalsToBackend();
        addNotesPageTree();
        string label_rule = "elephant";
        backend_.deleteNode(label_rule);
        checkMultipleDeletedNode(backend_, createElephantNodes());
        checkMultipleGetNode(backend_, createLionNodes());
        checkMultipleGetNode(backend_, createParrotNodes());
    }
}

void BackendTestbed::addAnimalsToBackend() {
    vector<TreeNode> lion_nodes = createLionNodes();
    backend_.upsertNode(lion_nodes);
    vector<TreeNode> elephant_nodes = createElephantNodes();
    backend_.upsertNode(elephant_nodes);
    vector<TreeNode> parrot_nodes = createParrotNodes();
    backend_.upsertNode(parrot_nodes);
}

void BackendTestbed::addNotesPageTree() {
    vector<TreeNode> notes = {createNotesPageTree()};
    backend_.upsertNode(notes);
}

void BackendTestbed::testAnimalNodesNoElephant(string label_prefix) {
    checkMultipleDeletedNode(backend_, prefixNodeLabels(label_prefix, createElephantNodes()));
    checkMultipleGetNode(backend_, prefixNodeLabels(label_prefix, createLionNodes()));
    checkMultipleGetNode(backend_, prefixNodeLabels(label_prefix, createParrotNodes()));
}

void BackendTestbed::testBackendLogically(string label_prefix) {
    checkMultipleGetNode(backend_, prefixNodeLabels(label_prefix, createLionNodes()));
    checkMultipleGetNode(backend_, prefixNodeLabels(label_prefix, createElephantNodes()));
    checkMultipleGetNode(backend_, prefixNodeLabels(label_prefix, createParrotNodes()));

    auto notes_pages = backend_.getPageTree(label_prefix + "notes");
    if (useCatch2) {
        REQUIRE(notes_pages == prefixNodeLabels(label_prefix, collectAllNotes()));
    }
    else {
        assert(notes_pages == prefixNodeLabels(label_prefix, collectAllNotes()));
    }

    {
        MemoryTree temp_tree;
        SimpleBackend temp_backend(temp_tree);
        auto full_tree = backend_.getFullTree();
        temp_backend.upsertNode(full_tree);
        checkMultipleGetNode(temp_backend, prefixNodeLabels(label_prefix, createElephantNodes()));
        checkMultipleGetNode(temp_backend, prefixNodeLabels(label_prefix, createLionNodes()));
        checkMultipleGetNode(temp_backend, prefixNodeLabels(label_prefix, createParrotNodes()));
        auto notes_pages = temp_backend.getPageTree(label_prefix + "notes");
        if (useCatch2) {
            REQUIRE(notes_pages == prefixNodeLabels(label_prefix, collectAllNotes()));
        }
        else {
            assert(notes_pages == prefixNodeLabels(label_prefix, collectAllNotes()));
        }
    }
    
    if (should_test_changes_) {
        auto label_rule = label_prefix + "elephant";
        backend_.deleteNode(label_rule);
        checkMultipleDeletedNode(backend_, prefixNodeLabels(label_prefix, createElephantNodes()));
        checkMultipleGetNode(backend_, prefixNodeLabels(label_prefix, createLionNodes()));
        checkMultipleGetNode(backend_, prefixNodeLabels(label_prefix, createParrotNodes()));
    }

    if (!should_test_notifications_ || !should_test_changes_) {
        return;
    }

    bool lion_node_created = false;
    bool lion_node_deleted = false;
    string lion_label = label_prefix + "lion";
    auto backend_address = &backend_;
    backend_.registerNodeListener("lion_listener", lion_label, false, 
        [lion_label, &lion_node_created, &lion_node_deleted, backend_address](Backend& notified_backend, const string label_rule, const fplus::maybe<TreeNode> node) {
        if (useCatch2) {
            REQUIRE(lion_label == label_rule);
            REQUIRE(backend_address == &notified_backend);
        }
        else {
            assert(lion_label == label_rule);
            assert(backend_address == &notified_backend);
        }
        if (node.is_just()) {
            auto found_node = node.get_with_default(TreeNode());
            if (found_node.getLabelRule() == lion_label) {
                lion_node_created = true;
            }
        } else {
            lion_node_deleted = true;
        }
    });
    auto prefixed_lion_nodes = prefixNodeLabels(label_prefix, createLionNodes());
    backend_.upsertNode(prefixed_lion_nodes);
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(lion_node_created);
    }
    else {
        assert(lion_node_created);
    }
    backend_.deleteNode(lion_label);
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(lion_node_deleted);
    }
    else {
        assert(lion_node_deleted);
    }
    backend_.deregisterNodeListener("lion_listener", lion_label);
    backend_.processNotifications();
    lion_node_created = false;
    lion_node_deleted = false;
    backend_.upsertNode(prefixed_lion_nodes);
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(!lion_node_created);
    }
    else {
        assert(!lion_node_created);
    }
    backend_.deleteNode(lion_label);
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(!lion_node_deleted);
    }
    else {
        assert(!lion_node_deleted);
    }
}

void BackendTestbed::testPeerNotification(Backend &to_be_modified, uint32_t notification_delay, string label_prefix)
{
    if(!should_test_notifications_ || !should_test_changes_) {
        return;
    }
    bool lion_node_created = false;
    bool lion_node_deleted = false;
    string lion_label = label_prefix + "lion";
    to_be_modified.deleteNode(lion_label);
    auto backend_address = &backend_;
    backend_.registerNodeListener("lion_listener", lion_label, false, 
        [lion_label, &lion_node_created, &lion_node_deleted, backend_address](Backend& notified_backend, const string label_rule, const fplus::maybe<TreeNode> node) {
        if (useCatch2) {
            REQUIRE(lion_label == label_rule);
            REQUIRE(backend_address == &notified_backend);
        }
        else {
            assert(lion_label == label_rule);
            assert(backend_address == &notified_backend);
        }
        if (node.is_just()) {
            auto found_node = node.get_with_default(TreeNode());
            if (found_node.getLabelRule() == lion_label) {
                lion_node_created = true;
            }
        } else {
            lion_node_deleted = true;
        }
    });
    auto prefixed_lion_nodes = prefixNodeLabels(label_prefix, createLionNodes());
    to_be_modified.upsertNode(prefixed_lion_nodes);
    to_be_modified.processNotifications();
    this_thread::sleep_for(chrono::milliseconds(notification_delay));
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(lion_node_created);
    }
    else {
        assert(lion_node_created);
    }
    to_be_modified.deleteNode(lion_label);
    to_be_modified.processNotifications();
    this_thread::sleep_for(chrono::milliseconds(notification_delay));
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(lion_node_deleted);
    }
    else {
        assert(lion_node_deleted);
    }
    backend_.deregisterNodeListener("lion_listener", lion_label);
    to_be_modified.processNotifications();
    this_thread::sleep_for(chrono::milliseconds(notification_delay));
    backend_.processNotifications();
    lion_node_created = false;
    lion_node_deleted = false;
    to_be_modified.upsertNode(prefixed_lion_nodes);
    to_be_modified.processNotifications();
    this_thread::sleep_for(chrono::milliseconds(notification_delay));
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(!lion_node_created);
    }
    else {
        assert(!lion_node_created);
    }
    to_be_modified.deleteNode(lion_label);
    to_be_modified.processNotifications();
    this_thread::sleep_for(chrono::milliseconds(notification_delay));
    backend_.processNotifications();
    if (useCatch2) {
        REQUIRE(!lion_node_deleted);
    }
    else {
        assert(!lion_node_deleted);
    }

}

File: js_client_lib/test/backend_testbed/backend_testbed.js
// JS Backend Testbed helpers for WWATP parity suite
// Ported from C++ backend_testbed.h
import { TreeNode, TreeNodeVersion } from '../../interface/tree_node.js';
import { Just, Nothing } from '../../interface/maybe.js';

// Helper: create a TreeNode with no content
export function createNoContentTreeNode(labelRule, description, propertyInfos, version, childNames, queryHowTo = Nothing, qaSequence = Nothing) {
  return new TreeNode({
    labelRule,
    description,
    propertyInfos,
    version: version instanceof TreeNodeVersion ? version : new TreeNodeVersion(version || {}),
    childNames,
    propertyData: new Uint8Array(0),
    queryHowTo,
    qaSequence,
  });
}

// Helper: create an "animal" node with property data
export function createAnimalNode(animal, description, propertyInfos, version, childNames, propertyDataStrings, queryHowTo, qaSequence) {
  // propertyDataStrings: array of string values for each property
  // propertyInfos: [{type, name}]
  let propertyData = buildPropertyData(propertyInfos, propertyDataStrings);
  return new TreeNode({
    labelRule: animal,
    description,
    propertyInfos,
    version: version instanceof TreeNodeVersion ? version : new TreeNodeVersion(version || {}),
    childNames,
    propertyData,
    // Treat empty strings as Nothing to match chunk-based decode semantics
    queryHowTo: (typeof queryHowTo === 'string' && queryHowTo.length === 0) ? Nothing : Just(queryHowTo),
    qaSequence: (typeof qaSequence === 'string' && qaSequence.length === 0) ? Nothing : Just(qaSequence),
  });
}

function buildPropertyData(propertyInfos, propertyDataStrings) {
  // Builds Uint8Array for propertyData from propertyInfos and propertyDataStrings
  // Only supports uint64 and string for now (as in C++ testbed)
  let totalLen = 0;
  let parts = [];
  for (let i = 0; i < propertyInfos.length; i++) {
    const { type } = propertyInfos[i];
    const value = propertyDataStrings[i];
    if (type === 'uint64') {
      // 8 bytes, little-endian
      const buf = new ArrayBuffer(8);
      const dv = new DataView(buf);
      dv.setBigUint64(0, BigInt(value), true);
      parts.push(new Uint8Array(buf));
      totalLen += 8;
    } else if (type === 'string') {
      // 4-byte length prefix + UTF-8 bytes
      const strBytes = new TextEncoder().encode(value);
      const lenBuf = new ArrayBuffer(4);
      new DataView(lenBuf).setUint32(0, strBytes.length, true);
      parts.push(new Uint8Array(lenBuf));
      parts.push(strBytes);
      totalLen += 4 + strBytes.length;
    }
    // Extend for other types as needed
  }
  let out = new Uint8Array(totalLen);
  let offset = 0;
  for (let p of parts) {
    out.set(p, offset);
    offset += p.length;
  }
  return out;
}

export function createAnimalDossiers(animalNode) {
  // For each child, create a dossier node and 3 note nodes
  let animal_dossiers = [];
  for (const child_name of animalNode.getChildNames()) {
    const child_label_rule = animalNode.getLabelRule() + '/' + child_name;
    const child_version = new TreeNodeVersion({ versionNumber: 1, maxVersionSequence: 256, policy: 'public', collisionDepth: Just(2) });
    const child_node = createNoContentTreeNode(child_label_rule, 'Animal Dossier', [], child_version, [], Nothing, Nothing);
    let note_names = [];
    for (let i = 0; i < 3; ++i) {
      const note_label_rule = child_label_rule + '/note' + i;
      note_names.push('note' + i);
      const note_version = new TreeNodeVersion({ versionNumber: 1, maxVersionSequence: 256, policy: 'public', collisionDepth: Just(2) });
      const note_node = createNoContentTreeNode(note_label_rule, child_name + ' note ' + i, [], note_version, [], Nothing, Nothing);
      animal_dossiers.push(note_node);
    }
    child_node.setChildNames(note_names);
    animal_dossiers.push(child_node);
  }
  return animal_dossiers;
}

export function createLionNodes() {
  const lion = createAnimalNode(
    'lion',
    'King of the jungle',
    [{ type: 'uint64', name: 'popularity' }, { type: 'string', name: 'diet' }],
    new TreeNodeVersion({ versionNumber: 1, maxVersionSequence: 256, policy: 'public', authors: Just('tester'), readers: Just('tester'), collisionDepth: Just(2) }),
    ['Simba', 'Nala'],
    ['10', 'carnivore'],
    'url duh!',
    'Zookeeper1: Lions are majestic.\nZookeeper2: Indeed, they are the kings of the jungle.'
  );
  let lion_nodes = createAnimalDossiers(lion);
  lion_nodes.unshift(lion);
  return lion_nodes;
}

export function createElephantNodes() {
  const elephant = createAnimalNode(
    'elephant',
    'Largest land animal',
    [{ type: 'uint64', name: 'popularity' }, { type: 'string', name: 'diet' }],
    new TreeNodeVersion({ versionNumber: 1, maxVersionSequence: 256, policy: 'public', collisionDepth: Just(2) }),
    ['Dumbo', 'Babar'],
    ['8', 'herbivore'],
    'url duh!',
    'Zookeeper1: Elephants are so strong.\nZookeeper2: And they have great memory!'
  );
  let elephant_nodes = createAnimalDossiers(elephant);
  elephant_nodes.unshift(elephant);
  return elephant_nodes;
}

export function createParrotNodes() {
  const parrot = createAnimalNode(
    'parrot',
    'Colorful bird',
    [{ type: 'uint64', name: 'popularity' }, { type: 'string', name: 'diet' }],
    new TreeNodeVersion({ versionNumber: 1, maxVersionSequence: 256, policy: 'public', collisionDepth: Just(2) }),
    ['Polly', 'Jerome'],
    ['7', 'omnivore'],
    'url duh!',
    'Zookeeper1: Parrots can mimic sounds.\nZookeeper2: Yes, they are very intelligent birds.'
  );
  let parrot_nodes = createAnimalDossiers(parrot);
  parrot_nodes.unshift(parrot);
  return parrot_nodes;
}

export function collectAllNotes() {
  const lion_nodes = createLionNodes();
  const elephant_nodes = createElephantNodes();
  const parrot_nodes = createParrotNodes();
  let all_notes = [];
  for (const node of [...lion_nodes, ...elephant_nodes, ...parrot_nodes]) {
    if (node.getLabelRule().includes('note')) {
      all_notes.push(node);
    }
  }
  return all_notes;
}

export function createNotesPageTree() {
  const noteNodes = collectAllNotes();
  const note_labels = noteNodes.map((n) => n.getLabelRule());
  // page_nodes property: variable-size, name 'page_nodes', value is JSON array of label rules
  const page_nodes_info = [{ type: 'string', name: 'page_nodes' }];
  const page_nodes_data = [JSON.stringify(note_labels)];
  const node = createNoContentTreeNode(
    'notes',
    'Animal Notes Page',
    page_nodes_info,
    new TreeNodeVersion({ versionNumber: 1, maxVersionSequence: 256, policy: 'public', collisionDepth: Just(2) }),
    note_labels,
    Nothing,
    Nothing
  );
  node.setPropertyData(buildPropertyData(page_nodes_info, page_nodes_data));
  return node;
}

export function prefixNodeLabels(labelPrefix, nodes) {
  return nodes.map((node) => {
    const n = node.clone();
    if (labelPrefix) {
      n.setLabelRule(labelPrefix + n.getLabelRule());
      n.setChildNames(n.getChildNames().map((c) => labelPrefix + c));
    }
    return n;
  });
}

// BackendTestbed class
export class BackendTestbed {
  constructor(backend, opts = {}) {
    this.backend = backend;
    this.shouldTestNotifications = opts.shouldTestNotifications ?? true;
    this.shouldTestChanges = opts.shouldTestChanges ?? true;
  }

  addAnimalsToBackend() {
    this.backend.upsertNode(createLionNodes());
    this.backend.upsertNode(createElephantNodes());
    this.backend.upsertNode(createParrotNodes());
  }

  addNotesPageTree() {
    this.backend.upsertNode([createNotesPageTree()]);
  }

  testBackendLogically(labelPrefix = '') {
    // Check all animal nodes exist
    checkMultipleGetNode(this.backend, prefixNodeLabels(labelPrefix, createLionNodes()));
    checkMultipleGetNode(this.backend, prefixNodeLabels(labelPrefix, createElephantNodes()));
    checkMultipleGetNode(this.backend, prefixNodeLabels(labelPrefix, createParrotNodes()));

    // Check notes page tree
    const notesPages = this.backend.getPageTree(labelPrefix + 'notes');
    const expectedNotes = prefixNodeLabels(labelPrefix, collectAllNotes());
    if (notesPages.length !== expectedNotes.length || !notesPages.every((n, i) => n.equals(expectedNotes[i]))) {
      throw new Error('Notes page tree does not match expected notes');
    }

    // Full tree round-trip
  this.backend.getFullTree();
    // For round-trip, create a new backend and upsert all nodes
    // (Assume SimpleBackend is available in test file)

    // Test changes: delete elephant, check
    if (this.shouldTestChanges) {
      const label_rule = labelPrefix + 'elephant';
      this.backend.deleteNode(label_rule);
      checkMultipleDeletedNode(this.backend, prefixNodeLabels(labelPrefix, createElephantNodes()));
      checkMultipleGetNode(this.backend, prefixNodeLabels(labelPrefix, createLionNodes()));
      checkMultipleGetNode(this.backend, prefixNodeLabels(labelPrefix, createParrotNodes()));
    }
    // Notification tests are handled in test file for now
  }

  // Port of C++ BackendTestbed::testPeerNotification
  // toBeModified: Backend performing the external mutations (peer)
  // notificationDelay: ms to wait between peer action and local processNotifications
  // labelPrefix: optional prefix for labels
  async testPeerNotification(toBeModified, notificationDelay = 0, labelPrefix = '') {
    if (!this.shouldTestNotifications || !this.shouldTestChanges) return;

    const sleep = (ms) => new Promise((r) => setTimeout(r, Math.max(0, Number(ms) | 0)));
    const awaitIfPromise = async (v) => (v && typeof v.then === 'function' ? await v : v);

    let lionCreated = false;
    let lionDeleted = false;
    const lionLabel = String(labelPrefix || '') + 'lion';

    // Ensure clean state on the peer backend
    await awaitIfPromise(toBeModified.deleteNode(lionLabel));

    // Register listener on our local backend
    const selfBackend = this.backend;
    await awaitIfPromise(this.backend.registerNodeListener('lion_listener', lionLabel, false, (notifiedBackend, labelRule, maybeNode) => {
      if (labelRule !== lionLabel) throw new Error(`Listener label mismatch: got ${labelRule}, expected ${lionLabel}`);
      if (notifiedBackend !== selfBackend) throw new Error('Listener backend instance mismatch');
      if (maybeNode && typeof maybeNode.isJust === 'function' && maybeNode.isJust()) {
        const node = maybeNode.getOrElse(null);
        if (node && typeof node.getLabelRule === 'function' && node.getLabelRule() === lionLabel) {
          lionCreated = true;
        }
      } else {
        // Nothing => deletion
        lionDeleted = true;
      }
    }));

    const prefixedLionNodes = prefixNodeLabels(labelPrefix || '', createLionNodes());

    // Create on the peer
    await awaitIfPromise(toBeModified.upsertNode(prefixedLionNodes));
    await awaitIfPromise(toBeModified.processNotifications());
    await sleep(notificationDelay);
    await awaitIfPromise(this.backend.processNotifications());
    if (!lionCreated) throw new Error('Expected lion node creation notification');

    // Now delete on the peer
    await awaitIfPromise(toBeModified.deleteNode(lionLabel));
    await awaitIfPromise(toBeModified.processNotifications());
    await sleep(notificationDelay);
    await awaitIfPromise(this.backend.processNotifications());
    if (!lionDeleted) throw new Error('Expected lion node deletion notification');

    // Deregister and ensure no further notifications are observed
    await awaitIfPromise(this.backend.deregisterNodeListener('lion_listener', lionLabel));
    await awaitIfPromise(toBeModified.processNotifications());
    await sleep(notificationDelay);
    await awaitIfPromise(this.backend.processNotifications());

    lionCreated = false;
    lionDeleted = false;

    await awaitIfPromise(toBeModified.upsertNode(prefixedLionNodes));
    await awaitIfPromise(toBeModified.processNotifications());
    await sleep(notificationDelay);
    await awaitIfPromise(this.backend.processNotifications());
    if (lionCreated) throw new Error('Listener should be deregistered; unexpected creation notification');

    await awaitIfPromise(toBeModified.deleteNode(lionLabel));
    await awaitIfPromise(toBeModified.processNotifications());
    await sleep(notificationDelay);
    await awaitIfPromise(this.backend.processNotifications());
    if (lionDeleted) throw new Error('Listener should be deregistered; unexpected deletion notification');
  }
}

// Test helpers
export function checkGetNode(backend, labelRule, expectedNode) {
  const maybeNode = backend.getNode(labelRule);
  if (maybeNode.isJust()) {
    const foundNode = maybeNode.getOrElse(null);
    if (!foundNode.equals(expectedNode)) {
      throw new Error(`Node mismatch for ${labelRule}`);
    }
  } else {
    throw new Error(`Node not found: ${labelRule}`);
  }
}

export function checkMultipleGetNode(backend, expectedNodes) {
  for (const expectedNode of expectedNodes) {
    checkGetNode(backend, expectedNode.getLabelRule(), expectedNode);
  }
}

export function checkDeletedNode(backend, labelRule) {
  const maybeNode = backend.getNode(labelRule);
  if (!maybeNode.isNothing()) {
    throw new Error(`Node should be deleted: ${labelRule}`);
  }
}

export function checkMultipleDeletedNode(backend, expectedNodes) {
  for (const expectedNode of expectedNodes) {
    checkDeletedNode(backend, expectedNode.getLabelRule());
  }
}
